knora/dsplib/utils/xml_upload.py

```diff
@@ -843,65 +769,147 @@ def xml_upload(input_file: str, server: str, user: str, password: str, imgdir: s
-        # because they do not have names
-        if not (
-            isinstance(elem, etree._Comment)
-            or isinstance(elem, etree._ProcessingInstruction)
-        ):
-            # Remove a namespace URI in the element's name
-            elem.tag = etree.QName(elem).localname
-
-    # Remove unused namespace declarations
-    etree.cleanup_namespaces(tree)
+        if not (isinstance(elem, etree._Comment) or isinstance(elem, etree._ProcessingInstruction)):
+            elem.tag = etree.QName(elem).localname  # remove namespace URI in the element's name
+    etree.cleanup_namespaces(tree)  # remove unused namespace declarations
 
     knora = tree.getroot()
     default_ontology = knora.attrib['default-ontology']
     shortcode = knora.attrib['shortcode']
 
+    resources: list[XMLResource] = []
+    permissions: dict[str, XmlPermission] = {}
     for child in knora:
-        # get all permissions
         if child.tag == "permissions":
             permission = XmlPermission(child, proj_context)
             permissions[permission.id] = permission
-        # get all resources
         elif child.tag == "resource":
             resources.append(XMLResource(child, default_ontology))
 
+    # get the project information and project ontology from the server
+    project = ResourceInstanceFactory(con, shortcode)
+    permissions_lookup: dict[str, Permissions] = {s: perm.get_permission_instance() for s, perm in permissions.items()}
+    resclass_name_2_type: dict[str, type] = {s: project.get_resclass_type(s) for s in project.get_resclass_names()}
+
     # temporarily remove circular references, but only if not an incremental upload
     if not incremental:
         resources, stashed_xml_texts, stashed_resptr_props = remove_circular_references(resources, verbose)
     else:
         stashed_xml_texts = dict()
         stashed_resptr_props = dict()
 
-    sipi_server = Sipi(sipi, con.get_token())
+    id2iri_mapping: dict[str, str] = {}
+    failed_uploads: list[str] = []
 
-    # get the project information and project ontology from the server
-    project = ResourceInstanceFactory(con, shortcode)
+    try:
+        id2iri_mapping, failed_uploads = upload_resources(verbose, resources, imgdir, sipi_server, permissions_lookup,
+                                                          resclass_name_2_type, id2iri_mapping, con, failed_uploads)
+    except BaseException as err:
+        handle_upload_error(err, input_file, id2iri_mapping, failed_uploads, stashed_xml_texts, stashed_resptr_props)
+        exit(1)
+
+    # update the resources with the stashed XML texts
+    nonapplied_xml_texts = {}
+    if len(stashed_xml_texts) > 0:
+        try:
+            nonapplied_xml_texts = update_stashed_xml_texts(verbose, id2iri_mapping, con, stashed_xml_texts)
+        except BaseException as err:
+            handle_upload_error(err, input_file, id2iri_mapping, failed_uploads, stashed_xml_texts, stashed_resptr_props)
+            exit(1)
```

Comment:
as above. Also, `except BaseException` is (practically) equivalent and imo less clear (and less common) than `except Exception`

Reply:
`Exception` is the base class for the non-system-exiting exceptions. But I want to catch the system-exiting ones too, so it's necessary to catch `BaseException`.

Reply:
Ok, then I don't understand why you would want to catch system-exiting exceptions at this point (which in my opinion should never be done without re-raising them right away)